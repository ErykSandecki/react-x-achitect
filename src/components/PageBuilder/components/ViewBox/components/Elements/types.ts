// types
import { MouseMode } from 'components/PageBuilder/enums';
import { TElement } from 'types';

export type TElementProps = {
  className: string;
  id: TElement['id'];
  mouseMode: MouseMode;
  parentId: TElement['parentId'];
  type: TElement['type'];
};
