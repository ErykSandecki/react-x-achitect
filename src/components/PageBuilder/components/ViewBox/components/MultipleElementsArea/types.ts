// types
import { TElement, TRectCoordinates } from 'types';

export type TElements = {
  elementsCordinates: Array<{
    coordinates: TRectCoordinates;
    id: TElement['id'];
  }>;
  outline: TRectCoordinates;
};
