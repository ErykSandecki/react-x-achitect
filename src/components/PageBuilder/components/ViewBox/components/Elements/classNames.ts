export const className = 'Elements';

export const classNames = {
  [className]: {
    name: className,
    modificators: {
      frameDraggable: `${className}--frame-draggable`,
    },
  },
};
