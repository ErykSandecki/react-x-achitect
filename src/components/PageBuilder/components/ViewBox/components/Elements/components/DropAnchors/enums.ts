export enum DropAnchorsPosition {
  top = 'top',
  bottom = 'bottom',
  left = 'left',
  right = 'right',
}
