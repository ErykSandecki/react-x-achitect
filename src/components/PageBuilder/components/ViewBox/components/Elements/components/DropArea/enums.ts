export enum DropAreaPosition {
  top = 'top',
  bottom = 'bottom',
  left = 'left',
  right = 'right',
}
