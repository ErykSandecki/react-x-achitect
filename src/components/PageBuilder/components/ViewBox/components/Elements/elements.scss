.Elements {
  &--frame-draggable {
    pointer-events: none;
  }
}
