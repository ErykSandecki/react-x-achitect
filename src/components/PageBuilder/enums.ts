export enum Constrain {
  bottom = 'bottom',
  centerHorizontal = 'centerHorizontal',
  centerVertical = 'centerVertical',
  left = 'left',
  right = 'right',
  top = 'top',
}
