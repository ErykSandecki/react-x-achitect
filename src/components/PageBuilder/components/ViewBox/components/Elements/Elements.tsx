import { FC, memo } from 'react';
import { useSelector } from 'react-redux';

// components
import DropAnchors from './components/DropAnchors/DropAnchors';
import Frame from './components/Frame/Frame';

// hooks
import { useTheme } from 'hooks';

// others
import { classNames } from './classNames';

// store
import { filtredStaticDataSelectorCreator } from 'store/pageBuilder/selectors';

// styles
import styles from './elements.scss';

// types
import { ElementType, TElement } from 'types';
import { MouseMode } from 'components/PageBuilder/enums';

export type TElementsProps = {
  eventsDisabled: boolean;
  isSelected?: boolean;
  mouseMode: MouseMode;
  parentId: TElement['parentId'];
};

const Elements: FC<TElementsProps> = ({
  eventsDisabled,
  isSelected = false,
  mouseMode,
  parentId,
}) => {
  const staticData = useSelector(filtredStaticDataSelectorCreator(parentId));
  const { classNamesWithTheme, cx } = useTheme(classNames, styles);

  return staticData.map(({ id, parentId, position, type }, index) => {
    const Wrapper =
      !isSelected && position === 'relative'
        ? ({ children }) => (
            <DropAnchors
              id={id}
              index={index}
              mouseMode={mouseMode}
              parentId={parentId}
            >
              {children}
            </DropAnchors>
          )
        : ({ children }) => <>{children}</>;

    switch (type) {
      case ElementType.frame:
        return (
          <Wrapper key={id}>
            <Frame
              className={cx(classNamesWithTheme.element.name, [
                classNamesWithTheme.element.modificators.eventsDisabled,
                eventsDisabled,
              ])}
              id={id}
              mouseMode={mouseMode}
              parentId={parentId}
              type={type}
            />
          </Wrapper>
        );
      default:
        return <></>;
    }
  });
};

export default memo(Elements);
