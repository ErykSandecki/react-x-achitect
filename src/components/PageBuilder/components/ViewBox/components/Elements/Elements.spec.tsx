import { Provider } from 'react-redux';
import { render } from '@testing-library/react';

// components
import Elements from './Elements';

// mocks
import {
  elementAllDataMock,
  elementDynamicDataMock,
  elementStaticDataMock,
  pageBuilderStateMock,
} from 'test/mocks/reducer/pageBuilderMock';
import { wholeStateMock } from 'test/mocks/reducer/wholeStateMock';
import 'test/mocks/sagas/allSagas';

// others
import { REDUCER_KEY as PAGE_BUILDER } from 'store/pageBuilder/actionsType';

// store
import { configureStore } from 'store/store';

// types
import { MouseMode } from 'components/PageBuilder/enums';
import { ElementType } from 'types';

const stateMock = {
  ...wholeStateMock,
  [PAGE_BUILDER]: {
    ...pageBuilderStateMock[PAGE_BUILDER],
  },
};

describe('Elements snapshots', () => {
  it('should render Elements', () => {
    // mock
    const store = configureStore(stateMock);

    // before
    const { asFragment } = render(
      <Provider store={store}>
        <Elements
          eventsDisabled={false}
          mouseMode={MouseMode.default}
          parentId="-1"
        />
      </Provider>,
    );

    // result
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render Element in relative position', () => {
    // mock
    const store = configureStore({
      [PAGE_BUILDER]: {
        ...pageBuilderStateMock[PAGE_BUILDER],
        elements: {
          allData: {
            ['-1']: {
              ...pageBuilderStateMock[PAGE_BUILDER].elements.allData['-1'],
              children: [elementAllDataMock.id],
            },
            [elementAllDataMock.id]: {
              ...elementAllDataMock,
              position: 'relative',
            },
          },
          dynamicData: {
            ['-1']: {
              ...pageBuilderStateMock[PAGE_BUILDER].elements.dynamicData['-1'],
              children: [elementDynamicDataMock.id],
            },
            [elementDynamicDataMock.id]: {
              ...elementDynamicDataMock,
              position: 'relative',
            },
          },
          staticData: {
            ['-1']: {
              ...pageBuilderStateMock[PAGE_BUILDER].elements.staticData['-1'],
              children: [elementStaticDataMock.id],
            },
            [elementStaticDataMock.id]: {
              ...elementStaticDataMock,
              position: 'relative',
            },
          },
        },
      },
    });

    // before
    const { asFragment } = render(
      <Provider store={store}>
        <Elements
          eventsDisabled={false}
          mouseMode={MouseMode.default}
          parentId="-1"
        />
      </Provider>,
    );

    // result
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render Frame', () => {
    // mock
    const store = configureStore({
      [PAGE_BUILDER]: {
        ...pageBuilderStateMock[PAGE_BUILDER],
        elements: {
          allData: {
            ['-1']: {
              ...pageBuilderStateMock[PAGE_BUILDER].elements.allData['-1'],
              children: [elementAllDataMock.id],
            },
            [elementAllDataMock.id]: elementAllDataMock,
          },
          dynamicData: {
            ['-1']: {
              ...pageBuilderStateMock[PAGE_BUILDER].elements.dynamicData['-1'],
              children: [elementDynamicDataMock.id],
            },
            [elementDynamicDataMock.id]: elementDynamicDataMock,
          },
          staticData: {
            ['-1']: {
              ...pageBuilderStateMock[PAGE_BUILDER].elements.staticData['-1'],
              children: [elementStaticDataMock.id],
            },
            [elementStaticDataMock.id]: elementStaticDataMock,
          },
        },
      },
    });

    // before
    const { asFragment } = render(
      <Provider store={store}>
        <Elements
          eventsDisabled={false}
          mouseMode={MouseMode.default}
          parentId="-1"
        />
      </Provider>,
    );

    // result
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render default Element', () => {
    // mock
    const store = configureStore({
      [PAGE_BUILDER]: {
        ...pageBuilderStateMock[PAGE_BUILDER],
        elements: {
          allData: {
            ['-1']: {
              ...pageBuilderStateMock[PAGE_BUILDER].elements.allData['-1'],
              children: [elementAllDataMock.id],
            },
            [elementAllDataMock.id]: {
              ...elementAllDataMock,
              type: ElementType.none,
            },
          },
          dynamicData: {
            ['-1']: {
              ...pageBuilderStateMock[PAGE_BUILDER].elements.dynamicData['-1'],
              children: [elementDynamicDataMock.id],
            },
            [elementDynamicDataMock.id]: elementDynamicDataMock,
          },
          staticData: {
            ['-1']: {
              ...pageBuilderStateMock[PAGE_BUILDER].elements.staticData['-1'],
              children: [elementStaticDataMock.id],
            },
            [elementStaticDataMock.id]: {
              ...elementStaticDataMock,
              type: ElementType.none,
            },
          },
        },
      },
    });

    // before
    const { asFragment } = render(
      <Provider store={store}>
        <Elements
          eventsDisabled={false}
          mouseMode={MouseMode.default}
          parentId="-1"
        />
      </Provider>,
    );

    // result
    expect(asFragment()).toMatchSnapshot();
  });
});
