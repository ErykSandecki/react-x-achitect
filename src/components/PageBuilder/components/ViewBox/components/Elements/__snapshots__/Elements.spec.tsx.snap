// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Elements snapshots should render Element in relative position 1`] = `
<DocumentFragment>
  <div
    class="DropAnchors DropAnchors--light"
  >
    <div
      class="css-q0izro Elements__element Elements__element--light Frame Frame--light Element Element--light"
      data-status="false"
      data-test-box=""
      id="1"
      style="background-color: rgb(255, 255, 255); height: 100px; left: 0px; position: relative; top: 0px; transform: rotate(0deg); width: 100px;"
    >
      <small
        class="css-q0izro Frame__label Frame__label--light Typography Typography--light Typography--neutral1 Typography--neutral1--light Typography--regular Typography--regular--light Typography--normal Typography--normal--light Typography--small Typography--small--light"
        data-test-text=""
        style="text-align: inherit;"
      >
        Frame
      </small>
    </div>
    <div
      class="DropAnchors__anchor DropAnchors__anchor--light DropAnchors__anchor--top DropAnchors__anchor--top--light"
      data-test-anchor="top"
    />
    <div
      class="DropAnchors__anchor DropAnchors__anchor--light DropAnchors__anchor--bottom DropAnchors__anchor--bottom--light"
      data-test-anchor="bottom"
    />
    <div
      class="DropAnchors__anchor DropAnchors__anchor--light DropAnchors__anchor--left DropAnchors__anchor--left--light"
      data-test-anchor="left"
    />
    <div
      class="DropAnchors__anchor DropAnchors__anchor--light DropAnchors__anchor--right DropAnchors__anchor--right--light"
      data-test-anchor="right"
    />
  </div>
</DocumentFragment>
`;

exports[`Elements snapshots should render Elements 1`] = `<DocumentFragment />`;

exports[`Elements snapshots should render Frame 1`] = `
<DocumentFragment>
  <div
    class="css-q0izro Elements__element Elements__element--light Frame Frame--light Element Element--light"
    data-status="false"
    data-test-box=""
    id="1"
    style="background-color: rgb(255, 255, 255); height: 100px; left: 0px; position: absolute; top: 0px; transform: rotate(0deg); width: 100px;"
  >
    <small
      class="css-q0izro Frame__label Frame__label--light Typography Typography--light Typography--neutral1 Typography--neutral1--light Typography--regular Typography--regular--light Typography--normal Typography--normal--light Typography--small Typography--small--light"
      data-test-text=""
      style="text-align: inherit;"
    >
      Frame
    </small>
  </div>
</DocumentFragment>
`;

exports[`Elements snapshots should render default Element 1`] = `<DocumentFragment />`;
